### ISSUE TITLE
Cron service

### ISSUE THREAD

**Issue by adamdyson** [WPN-XM/WPN-XM#537]
One thing I really miss since moving away from UniForm Server, is the ability to run cron jobs or scheduled tasks. I understand this can already be done natively in Windows, but having it built into the server control panel would be really useful.

I've done some digging and found the [source code](http://sourceforge.net/projects/miniserver/files/Uniform%20Server%20ZeroXI/ZeroXImodules/Source/UniController/UniController_1_1_9_src.exe/download) which you may be able to use or re-engineer to better suit your needs. The file that will be of interest is called `cron_form.pas` and seems pretty straight forward. When you get a chance to look the code, you might notice that they support running task in seconds, which I've utilized on a couple of projects when you almost need a daemon process (something I'm scared to attempt on Windows with PHP).

There are also these Wiki articles explaining its usage [1](http://wiki.uniformserver.com/index.php/Coral:_cron_intro), [2](http://wiki.uniformserver.com/index.php/Coral:_cron_configuration_detail), [3](http://wiki.uniformserver.com/index.php/Coral:_cron_portable_design).

**Comment by jakoch** [comment]
Thank you for the feature suggestion.

I think a basic abstraction on-top of the native "schtasks" command should provide the functionality in a more stable and more standardized way. 
(Compared to registering a service, which uses "ping" or "unidelay" to get the time signature 
or re-implementing the run_cron() procedure from their source). 

While the accepted arguments of "schtasks" do not indicate support for SECONDS (https://msdn.microsoft.com/en-us/library/windows/desktop/bb736357(v=vs.85).aspx), 
the XML file seems to support them. http://stackoverflow.com/a/18278262/1163786

If that is true, we could build a simple form to define the new task, 
save it as XML and register it to the task schedular by using 
`schtasks.exe /Create /XML task.xml /tn taskname`

Todo:
- [ ] ability to run cron jobs or scheduled tasks
- [ ] cronjob entry/definition must support "seconds"

**Comment by adamdyson** [comment]
It's been a while since I last looked into the Windows task scheduler, but I have a feeling executing tasks in seconds is not possible. While the command line arguments and XML file might accept a seconds value, it will be ignored and rounded down. I'm not 100% sure though.

If I remember correctly, the only way around it is to create multiple triggers with an offset. Looking at this [page](https://msdn.microsoft.com/en-us/library/windows/desktop/aa383619(v=vs.85).aspx) it appears there are two API's v1.0 & v2.0, each having different ways of defining them, which could be a problem?

Couldn't this be built into the Server Control Panel without having to register an additional service? IMO having the tasks defined in a file (like cron) is better, and makes it more portable.

**Comment by jakoch** [comment]
> but I have a feeling executing tasks in seconds is not possible

Its possible to fire an event at a certain second (time-based triggered event),
but the smallest repetition interval is "1 minute". 

This fails:
```
<Triggers>
    <TimeTrigger>
      <Repetition>
        <Interval>PT0M30S</Interval>
        <Duration>P1D</Duration>
        <StopAtDurationEnd>false</StopAtDurationEnd>
      </Repetition>
      <StartBoundary>2016-01-01T03:21:12</StartBoundary>
      <Enabled>true</Enabled>
    </TimeTrigger>
```

https://msdn.microsoft.com/en-us/library/windows/desktop/aa382119(v=vs.85).aspx
A repetition interval with seconds, e.g. `PT10S` or `PT0M30S`, triggers an interval parsing error.

You are right! This will only work with multiple triggers to cover all the seconds.

> Couldn't this be built into the Server Control Panel without having to register an additional service?

Yes, thats possible. 
We could decide to use the cron-syntax, then parse the entries, calculate their trigger times and register QTimer events. I need to take a closer look at QTimer (http://doc.qt.io/qt-5.5/qtimer.html) and QTimerEvents. But after a quick glance, it seems that events fire in repetition interval defined by setInterval(), unless SingleShot() is true. That gives us one time-events and repeated events in milliseconds.

**Comment by adamdyson** [comment]
Yes that is odd but I'm not surprised. It's nice to know this type of scheduling is possible using QT. 

UniServer solves this problem of executing tasks in seconds using there own syntax. The `period` option takes the values: hourly, daily, weekly, monthly or numeric in seconds.

You could choose to follow the [quartz](http://www.quartz-scheduler.org/documentation/quartz-1.x/tutorials/crontrigger) syntax which supports seconds however the unix cron job syntax is more widely used. Even though seconds are not supported/definable in crontab, it may not matter.

```
* * * * * *
| | | | | | 
| | | | | +-- Year              (range: 1900-3000)
| | | | +---- Day of the Week   (range: 1-7, 1 standing for Monday)
| | | +------ Month of the Year (range: 1-12)
| | +-------- Day of the Month  (range: 1-31)
| +---------- Hour              (range: 0-23)
+------------ Minute            (range: 0-59)
```

If a developer chooses to execute a task (script or URI) using the expression `* * * * * *` this would normally get invoked every minute. But if the QTimer timeout has a quantum set to 1000ms that task would get executed every second, because it's basically saying "execute me on every cycle".

It might be worth setting a default timeout value of 60000 and give the user the option to change it by means of a text input, if they need more frequent scheduling? Handling race conditions should be up to the developer.

Reading back over everything I'm not sure what the best solution is, using a simpler syntax like UniServer makes it quite trivial to implement programmatically. E.g. you don't need to develop an expression parser but maybe there's already one freely available?
